File: src/test/java/com/tabnine/binary/fetch/LocalBinaryVersionsTests.java
package com.tabnine.binary.fetch;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.api.io.TempDir;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.junit.jupiter.MockitoExtension;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static com.tabnine.general.StaticConfig.*;
import static com.tabnine.testutils.TestData.*;
import static java.util.Collections.emptyList;
import static org.hamcrest.Matchers.empty;
import static org.hamcrest.Matchers.equalTo;
import static org.junit.Assert.assertThat;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
public class LocalBinaryVersionsTests {
    @TempDir
    public Path temporaryFolder;
    @Mock
    private BinaryValidator binaryValidator;
    @InjectMocks
    private LocalBinaryVersions localBinaryVersions;

    private String originalHome = System.getProperty(USER_HOME_PATH_PROPERTY);

    @BeforeEach
    public void setUp() {
        System.setProperty(USER_HOME_PATH_PROPERTY, temporaryFolder.toString());
    }

    @AfterEach
    public void tearDown() throws Exception {
        System.setProperty(USER_HOME_PATH_PROPERTY, originalHome);
    }

    @Test
    public void givenNoVersionsInTabNineBinaryFolderWhenListExistingThenResultIsEmptyList() throws Exception {
        Paths.get(temporaryFolder.toFile().toString(), TABNINE_FOLDER_NAME).toFile().mkdirs();

        assertThat(localBinaryVersions.listExisting(), equalTo(emptyList()));
    }

    @Test
    public void givenSomeValidLocalVersionsWhenListExistingVersionsThenResultContainsSortedValidVersions() throws Exception {
        when(binaryValidator.isWorking(versionFullPath(PRELATEST_VERSION))).thenReturn(true);
        when(binaryValidator.isWorking(versionFullPath(LATEST_VERSION))).thenReturn(true);
        when(binaryValidator.isWorking(versionFullPath(A_VERSION))).thenReturn(true);
        when(binaryValidator.isWorking(versionFullPath(ANOTHER_VERSION))).thenReturn(false);
        Stream.of(ANOTHER_VERSION, PRELATEST_VERSION, LATEST_VERSION, A_VERSION).map(version -> Paths.get(temporaryFolder.toString(), TABNINE_FOLDER_NAME, version).toFile()).forEach(File::mkdirs);


        assertThat(localBinaryVersions.listExisting(), equalTo(versions(LATEST_VERSION, PRELATEST_VERSION, A_VERSION)));
    }

    @Test
    public void givenAllLocalVersionsInvalidWhenListExistingVersionsThenResultIsEmptyList() throws Exception {
        when(binaryValidator.isWorking(Mockito.any())).thenReturn(false);
        Stream.of(A_VERSION, ANOTHER_VERSION).map(version -> Paths.get(temporaryFolder.toString(), TABNINE_FOLDER_NAME, version).toFile()).forEach(File::mkdirs);


        assertThat(localBinaryVersions.listExisting(), empty());
    }

    @Test
    public void givenNoActiveFileThenReturnNone() {
        assertThat(localBinaryVersions.activeVersion(), equalTo(Optional.empty()));
    }

    @Test
    public void givenActiveFileWithNoLinesThenReturnNone() throws IOException {
        Paths.get(temporaryFolder.toString(), TABNINE_FOLDER_NAME).toFile().mkdirs();
        Paths.get(temporaryFolder.toString(), TABNINE_FOLDER_NAME, ".active").toFile().createNewFile();
        assertThat(localBinaryVersions.activeVersion(), equalTo(Optional.empty()));
    }

    @Test
    public void givenActiveFileWithInvalidVersionThenReturnNone() throws IOException {
        Paths.get(temporaryFolder.toString(), TABNINE_FOLDER_NAME).toFile().mkdirs();
        Files.write(Paths.get(temporaryFolder.toString(), TABNINE_FOLDER_NAME, ".active"),  Stream.of(PREFERRED_VERSION).collect(Collectors.toList()));

        when(binaryValidator.isWorking(versionFullPath(PREFERRED_VERSION))).thenReturn(false);

        assertThat(localBinaryVersions.activeVersion(), equalTo(Optional.empty()));
    }

    @Test
    public void givenActiveFileWithValidVersionThenReturnVersion() throws IOException {
        Paths.get(temporaryFolder.toString(), TABNINE_FOLDER_NAME).toFile().mkdirs();
        Files.write(Paths.get(temporaryFolder.toString(), TABNINE_FOLDER_NAME, ".active"),  Stream.of(PREFERRED_VERSION).collect(Collectors.toList()));

        when(binaryValidator.isWorking(versionFullPath(PREFERRED_VERSION))).thenReturn(true);

        assertThat(localBinaryVersions.activeVersion(), equalTo(Optional.of(new BinaryVersion(PREFERRED_VERSION))));
    }

}

File: src/test/java/com/tabnine/binary/fetch/LocalBinaryVersionsTests.kt
package com.tabnine.binary.fetch

import com.tabnine.general.StaticConfig.* // ktlint-disable no-wildcard-imports
import com.tabnine.testUtils.TestData.* // ktlint-disable no-wildcard-imports
import io.mockk.every
import io.mockk.impl.annotations.InjectMockKs
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import org.hamcrest.Matchers.empty
import org.hamcrest.Matchers.equalTo
import org.junit.Assert.assertThat
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.junit.jupiter.api.io.TempDir
import java.io.File
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.util.Optional
import java.util.stream.Collectors.toList
import java.util.stream.Stream

@ExtendWith(MockKExtension::class)
class LocalBinaryVersionsTests {
    @TempDir
    @JvmField
    var temporaryFolder: Path? = null

    @MockK
    lateinit var binaryValidator: BinaryValidator

    @InjectMockKs
    lateinit var localBinaryVersions: LocalBinaryVersions

    private val originalHome = System.getProperty(USER_HOME_PATH_PROPERTY)

    @BeforeEach
    fun setup() {
        System.setProperty(USER_HOME_PATH_PROPERTY, temporaryFolder.toString())
    }

    @AfterEach
    fun teardown() {
        System.setProperty(USER_HOME_PATH_PROPERTY, originalHome)
    }

    @Test
    fun `given no versions in Tabnine binary folder when list existing then result is empty list`() {
        Paths.get(temporaryFolder!!.toFile().toString(), TABNINE_FOLDER_NAME).toFile().mkdirs()
        assertThat(localBinaryVersions.listExisting(), equalTo(emptyList<Any>()))
    }

    @Test
    fun `given some valid local versions when list existing versions then result contains sorted valid versions`() {
        every { binaryValidator.isWorking(versionFullPath(PRELATEST_VERSION)) } returns true
        every { binaryValidator.isWorking(versionFullPath(LATEST_VERSION)) } returns true
        every { binaryValidator.isWorking(versionFullPath(A_VERSION)) } returns true
        every { binaryValidator.isWorking(versionFullPath(ANOTHER_VERSION)) } returns false
        Stream.of(ANOTHER_VERSION, PRELATEST_VERSION, LATEST_VERSION, A_VERSION)
            .map { version: String? ->
                Paths.get(temporaryFolder.toString(), TABNINE_FOLDER_NAME, version).toFile()
            }
            .forEach { obj: File -> obj.mkdirs() }

        assertThat(
            localBinaryVersions.listExisting(),
            equalTo(versions(LATEST_VERSION, PRELATEST_VERSION, A_VERSION))
        )
    }

    @Test
    fun `given all local versions invalid when list existing versions then result is empty list`() {
        every { binaryValidator.isWorking(any()) } returns false
        Stream.of(A_VERSION, ANOTHER_VERSION).map {
            Paths.get(temporaryFolder.toString(), TABNINE_FOLDER_NAME, it).toFile()
        }.forEach { it.mkdirs() }

        assertThat(localBinaryVersions.listExisting(), empty())
    }

    @Test
    fun `given no active file then return none`() {
        assertThat(localBinaryVersions.activeVersion(), equalTo(Optional.empty<Any>()))
    }

    @Test
    fun `given active file with no lines then return none`() {
        Paths.get(temporaryFolder.toString(), TABNINE_FOLDER_NAME).toFile().mkdirs()
        Paths.get(temporaryFolder.toString(), TABNINE_FOLDER_NAME, ".active").toFile().createNewFile()

        assertThat(localBinaryVersions.activeVersion(), equalTo(Optional.empty<Any>()))
    }

    @Test
    fun `given active file with invalid version then return none`() {
        Paths.get(temporaryFolder.toString(), TABNINE_FOLDER_NAME).toFile().mkdirs()
        Files.write(
            Paths.get(temporaryFolder.toString(), TABNINE_FOLDER_NAME, ".active"),
            Stream.of(
                PREFERRED_VERSION
            ).collect(toList())
        )
        every { binaryValidator.isWorking(versionFullPath(PREFERRED_VERSION)) } returns false

        assertThat(localBinaryVersions.activeVersion(), equalTo(Optional.empty<Any>()))
    }

    @Test
    fun `given active file with valid version then return version`() {
        Paths.get(temporaryFolder.toString(), TABNINE_FOLDER_NAME).toFile().mkdirs()
        Files.write(
            Paths.get(temporaryFolder.toString(), TABNINE_FOLDER_NAME, ".active"),
            Stream.of(
                PREFERRED_VERSION
            ).collect(toList())
        )
        every { binaryValidator.isWorking(versionFullPath(PREFERRED_VERSION)) } returns true

        assertThat(
            localBinaryVersions.activeVersion(),
            equalTo(Optional.of(BinaryVersion(PREFERRED_VERSION)))
        )
    }
}
